timeinterval/timeinterval.go

```diff
@@ -0,0 +1,473 @@
+	}
+	if r.Begin == r.End {
+		return interface{}(beginStr), nil
+	}
+	endStr, ok := daysOfWeekInv[r.End]
+	if !ok {
+		return nil, fmt.Errorf("unable to convert %d into weekday string", r.End)
+	}
+	rangeStr := fmt.Sprintf("%s:%s", beginStr, endStr)
+	return interface{}(rangeStr), nil
+}
+
+//MarshalYAML implements the yaml.Marshaler interface for TimeRange
+func (tr TimeRange) MarshalYAML() (out interface{}, err error) {
+	startHr := tr.StartMinute / 60
+	endHr := tr.EndMinute / 60
+	startMin := tr.StartMinute % 60
+	endMin := tr.EndMinute % 60
+
+	startStr := fmt.Sprintf("%02d:%02d", startHr, startMin)
+	endStr := fmt.Sprintf("%02d:%02d", endHr, endMin)
+
+	yTr := yamlTimeRange{startStr, endStr}
+	return interface{}(yTr), err
+}
+
+//MarshalYAML implements the yaml.Marshaler interface for InclusiveRange
+func (ir InclusiveRange) MarshalYAML() (interface{}, error) {
+	if ir.Begin == ir.End {
+		return strconv.Itoa(ir.Begin), nil
+	}
+	out := fmt.Sprintf("%d:%d", ir.Begin, ir.End)
+	return interface{}(out), nil
+}
+
+// TimeLayout specifies the layout to be used in time.Parse() calls for time intervals
+const TimeLayout = "15:04"
+
+var validTime string = "^((([01][0-9])|(2[0-3])):[0-5][0-9])$|(^24:00$)"
+var validTimeRE *regexp.Regexp = regexp.MustCompile(validTime)
+
+// Given a time, determines the number of days in the month that time occurs in.
+func daysInMonth(t time.Time) int {
+	monthStart := time.Date(t.Year(), t.Month(), 1, 0, 0, 0, 0, t.Location())
+	monthEnd := monthStart.AddDate(0, 1, 0)
+	diff := monthEnd.Sub(monthStart)
+	return int(diff.Hours() / 24)
+}
+
+func clamp(n, min, max int) int {
+	if n <= min {
+		return min
+	}
+	if n >= max {
+		return max
+	}
+	return n
+}
+
+// ContainsTime returns true if the TimeInterval contains the given time, otherwise returns false
+func (tp TimeInterval) ContainsTime(t time.Time) bool {
+	if tp.Times != nil {
+		in := false
+		for _, validMinutes := range tp.Times {
+			if (t.Hour()*60+t.Minute()) >= validMinutes.StartMinute && (t.Hour()*60+t.Minute()) < validMinutes.EndMinute {
+				in = true
+				break
+			}
+		}
+		if !in {
+			return false
+		}
+	}
+	if tp.DaysOfMonth != nil {
+		in := false
+		for _, validDates := range tp.DaysOfMonth {
+			var begin, end int
+			daysInMonth := daysInMonth(t)
+			if validDates.Begin < 0 {
+				begin = daysInMonth + validDates.Begin + 1
+			} else {
+				begin = validDates.Begin
+			}
+			if validDates.End < 0 {
+				end = daysInMonth + validDates.End + 1
+			} else {
+				end = validDates.End
+			}
+			// Skip clamping if the beginning date is after the end of the month
+			if begin > daysInMonth {
+				continue
+			}
+			// Clamp to the boundaries of the month to prevent crossing into other months
+			begin = clamp(begin, -1*daysInMonth, daysInMonth)
+			end = clamp(end, -1*daysInMonth, daysInMonth)
+			if t.Day() >= begin && t.Day() <= end {
+				in = true
+				break
+			}
+		}
+		if !in {
```

Comment:
We do not need this thanks to break, we do not need in at all

Reply:
How do we avoid this? We need to loop over all the days/months/years etc to check if at least one of them matches, and return false if it doesn't.

Reply:
We need the return false

But if we break we will not need the `if`

Reply:
sorry I did speak too fast :)

Reply:
No problem, I felt this was clunky as well initially but I couldn't see a better way :) Cheers